### PR TITLE
Devicecrafting: imagine guncrafting but for literally anything

### PR DESCRIPTION
**Pull request by Iamgoofball**
Adds device crafting:
1. Print a Holder(handheld/structure)
2. Print a Trigger
3. Print 2-5 components
4. Hit Holder with triggers and components
5. Activate trigger, get effects

### THE PROBLEM:
I don't know what components to add. You guys should come up with components.

The framework works, I just don't know what to pack it with. Post them here or something idk lol

### CODE REVIEW

**Comment by juanototo2** [comment]
If components doesn't include the ability to make basic logic gates, rigging something like a welding tank to explode with a proximity sensor, or a death alarm using a radio I will be very dissapointed.

**Comment by GunHog** [comment]
# RAPID PROTOTYPING!!!
(Posted in Gooftext because this is a Goof PR)

**Comment by Iamgoofball** [comment]
yeah p. much rapid prototyping

File: code/game/objects/items.dm [hunk 1]
```diff
@@ -340,7 +340,29 @@ GLOBAL_VAR_INIT(rpg_loot_items, FALSE)
 
 			else if(S.can_be_inserted(src))
 				S.handle_item_insertion(src)
-
+	if(devicecrafting_holder)
+		if(istype(W, /obj/item/devicecrafting/device))
+			var/obj/item/devicecrafting/device/D = W
+			if(devicecrafting_holder.devices.len == devicecrafting_holder.max_devices)
```
Comment (MrPerson) [inline code/game/objects/items.dm]: >= would be better because of admins and bugs (which are the same thing really)

Comment (Iamgoofball) [inline code/game/objects/items.dm]: fair point

**Review by Cyberboss — DISMISSED** [review]
>deleted tools submodule

**Comment by Iamgoofball** [comment]
wait what

**Comment by ike709** [comment]
So this is basically integrated electronics from bay/polaris?

**Comment by kevinz000** [comment]
>any item can be used >not just a special itrm

...

**Comment by Iamgoofball** [comment]
Yeah

**Comment by Anonmare** [comment]
This is actually something I've really wanted lately.

File: code/modules/research/devicecrafting/device.dm [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+	for(var/D in my_holder.devices)
+		var/obj/item/devicecrafting/device/DE = D
+		DE.trigger(list("target" = target), src)
+	return
+
+/obj/item/devicecrafting/device/trigger/ranged // ranged attack
+	name = "beam trigger"
+	desc = "Can be used to activate onto a target from a range." // yeah, i see that, your codebase gave you good advice
+
+/obj/item/devicecrafting/device/trigger/ranged/on_afterattack(atom/target, mob/user, proximity_flag)
```
Comment (kevinz000) [inline code/modules/research/devicecrafting/device.dm]: where's the mouse params

Comment (Iamgoofball) [inline code/modules/research/devicecrafting/device.dm]: i suppose I could pass those along too but For What Purpose

**Comment by fludd12** [comment]
I can think of a very large amount of potentially impossible to code module ideas, if you'd like.

Would it be possible to, instead of each device having a hard cap of 5, have different holders which scale part limit with weight class?

**Comment by dannno** [comment]
I'm having trouble understanding just what this would implicate

could you provide some examples of potential devices

File: code/_onclick/item_attack.dm [hunk 3]
```diff
@@ -19,6 +21,29 @@
 	return
 
 /obj/attackby(obj/item/I, mob/living/user, params)
+	if(devicecrafting_holder)
+		if(istype(I, /obj/item/devicecrafting/device))
+			var/obj/item/devicecrafting/device/D = I
+			if(devicecrafting_holder.devices.len == devicecrafting_holder.max_devices)
+				user << "You can't add any more devices to this frame!"
```
Comment (RemieRichards) [inline code/_onclick/item_attack.dm]: ```to_chat```, ditto elsewhere.

File: code/modules/research/devicecrafting/device.dm [hunk 4]
```diff
@@ -0,0 +1,103 @@
+/obj/item/devicecrafting/device
+	name = "device"
+	icon = 'icons/obj/devicecrafting.dmi'
+	icon_state = "device"
+	var/obj/my_obj
+	var/datum/devicecrafting/holder/my_holder
+
+/obj/item/devicecrafting/device/proc/trigger(var/list/trigger_data, trigger)
```
Comment (RemieRichards) [inline code/modules/research/devicecrafting/device.dm]: no ```var/```

File: code/modules/research/devicecrafting/device.dm [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+/obj/item/devicecrafting/device/trigger/tesla // tiles surrouding my tile
+	name = "tesla trigger"
+	desc = "Activates when zapped by a tesla bolt." // oh shit i'm sorry
+
+/obj/item/devicecrafting/device/trigger/tesla/on_tesla(var/power)
+	for(var/D in my_holder.devices)
+		var/obj/item/devicecrafting/device/DE = D
+		DE.trigger(list("power" = power), src)
+	return
```
Comment (RemieRichards) [inline code/modules/research/devicecrafting/device.dm]: no trailing returns.

File: code/modules/research/devicecrafting/device.dm [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+/obj/item/devicecrafting/device/trigger/throwing // colliding
+	name = "impact trigger"
+	desc = "Activates when colliding after a throw." // sorry for what? our codebase taught us not to be ashamed of our line limits, 'specially since they're such good size and all
```
Comment (RemieRichards) [inline code/modules/research/devicecrafting/device.dm]: no.

**Comment by Jalleo** [comment]
Okay how about this one Goofball make the RCD RPD RLD and ARCD based on using this system. This would mean we dont have to copypaste as much with them.

File: code/modules/research/devicecrafting/device.dm (hunk 2 above)
Comment (RemieRichards) [inline code/modules/research/devicecrafting/device.dm]: pixel specific interactions. the exact pixels clicked are in those params.

File: code/modules/research/devicecrafting/device.dm [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+/obj/item/devicecrafting/device/trigger/ranged/on_afterattack(atom/target, mob/user, proximity_flag)
+	for(var/D in my_holder.devices)
+		var/obj/item/devicecrafting/device/DE = D
+		DE.trigger(list("target" = target, "user" = user, "prox_flag" = proximity_flag), src)
+	return
+
+/obj/item/devicecrafting/device/trigger/melee // melee attack
+	name = "blunt trigger"
+	desc = "Can be used to activate on a target by hitting them." // it gets bigger when i pull it, and sometimes, i pull on it so hard, i get a merge conflict
```
Comment (RemieRichards) [inline code/modules/research/devicecrafting/device.dm]: no, etc.

**Comment by Alek2ander** [comment]
Well easiest thing to start with is porting existing attachable devices to this: timers, remote signallers, prox sensors, infra sensors, igniters, grenades, TTVs, etc.

**Comment by Cyberboss** [comment]
@Iamgoofball Run the script in #28871 to fix the submodule thing

**Comment by Tacolizard** [comment]
how about using this for GUNCRAFTING???!!!!

**Comment by GunHog** [comment]
> how about using this for GUNCRAFTING???!!!!

Goof thinks ahead. This has probably been his plan all along.

**Comment by KorPhaeron** [comment]
>YOU CAN DO ANYTHING

>The Problem: You can't do anything

Sounds like we're ready for early access

**Comment by Cyberboss** [comment]
Merge upstream @Iamgoofball

**Comment by dannno** [comment]
I still don't understand the scope of this

**Comment by InsaneHyena** [comment]
We goonstation now!

**Comment by GunHog** [comment]
Goon has this?

**Comment by Anonmare** [comment]
@GunHog 
Sort of, Goon has the electrician job who can build complex devices, but they also call it mechcomp and it uses things like logic gates

https://wiki.ss13.co/MechComp

File: code/modules/research/devicecrafting/device.dm (hunk 6 above)
Comment (Armhulen) [inline code/modules/research/devicecrafting/device.dm]: do NOT remove that comment

File: code/modules/research/devicecrafting/device.dm (hunk 7 above)
Comment (Armhulen) [inline code/modules/research/devicecrafting/device.dm]: Armhulen 14 hours from now
do NOT remove that comment

Comment (Armhulen) [inline code/modules/research/devicecrafting/device.dm]: my computer's clock is broken :(

**Comment by KorPhaeron** [comment]
There is no way goof is going to finish this and also AI and also dwarves all at once

**Pull request closed without merging**